File: src/Core/LotMatching/Services/TradeAssociationTracker.py
from dataclasses import dataclass

from src.Core.FinancialEvents.Schemas.ProcessedGenericFormats import GenericTradeEvent


@dataclass
class TradeAssociation:
    Quantity: float
    Trade: GenericTradeEvent


class TradeAssociationTracker:
    tradeAcquiredTracker: dict[GenericTradeEvent, TradeAssociation] = dict()
    tradeSoldTracker: dict[GenericTradeEvent, TradeAssociation] = dict()

    def associateAcquiredTrade(self, event: GenericTradeEvent):
        if event not in self.tradeAcquiredTracker:
            self.tradeAcquiredTracker[event] = TradeAssociation(0, event)

    def associateSoldTrade(self, event: GenericTradeEvent):
        if event not in self.tradeSoldTracker:
            self.tradeSoldTracker[event] = TradeAssociation(0, event)

    def trackAcquiredQuantity(self, event: GenericTradeEvent, quantity: float):
        self.associateAcquiredTrade(event)

        tracker = self.tradeAcquiredTracker.get(event)
        if tracker is None:
            raise AssertionError("There is no tracker for the event ({})".format(event.ID))

        if tracker.Quantity + quantity > event.ExchangedMoney.UnderlyingQuantity:
            raise AssertionError(
                "Adding tracking quantity ({}) to event ({}) exceeds total quantity ({}) of the event".format(
                    quantity, event.ID, event.ExchangedMoney.UnderlyingQuantity
                )
            )

        tracker.Quantity += quantity

    def trackSoldQuantity(self, event: GenericTradeEvent, quantity: float):
        self.associateSoldTrade(event)

        tracker = self.tradeSoldTracker.get(event)
        if tracker is None:
            raise AssertionError("There is no tracker for the event ({})".format(event.ID))

        if tracker.Quantity + quantity > event.ExchangedMoney.UnderlyingQuantity:
            raise AssertionError(
                "Adding tracking quantity ({}) to event ({}) exceeds total quantity ({}) of the event".format(
                    quantity, event.ID, event.ExchangedMoney.UnderlyingQuantity
                )
            )

        tracker.Quantity += quantity

File: lib/core/lotMatching/TradeAssociationTracker.ts
import type { Trade } from "./Trade.ts";

export type TradeAssociation = {
  quantity: number;
  associatedTrade: Trade;
};

export class TradeAssociationTracker {
  private tradeAcquiredTracker: Map<string, TradeAssociation> = new Map();
  private tradeSoldTracker: Map<string, TradeAssociation> = new Map();

  private associateAcquiredTrade(event: Trade): void {
    if (!this.tradeAcquiredTracker.has(event.id)) {
      this.tradeAcquiredTracker.set(event.id, { quantity: 0, associatedTrade: event });
    }
  }

  private associateSoldTrade(event: Trade): void {
    if (!this.tradeSoldTracker.has(event.id)) {
      this.tradeSoldTracker.set(event.id, { quantity: 0, associatedTrade: event });
    }
  }

  trackTrade(event: Trade): void {
    if (event.quantity >= 0) {
      this.associateAcquiredTrade(event);
    } else {
      this.associateSoldTrade(event);
    }
  }

  trackAcquiredQuantity(event: Trade, quantity: number): void {
    this.associateAcquiredTrade(event);

    const tracker = this.tradeAcquiredTracker.get(event.id);
    if (tracker === undefined) {
      throw new Error(`There is no tracker for the event (${event.id})`);
    }

    if (tracker.quantity + quantity > event.quantity) {
      throw new Error(
        `Adding tracking quantity (${quantity}) to event (${event.id}) exceeds total quantity (${tracker.quantity + quantity} / ${event.quantity}) of the event`,
      );
    }

    tracker.quantity += quantity;
  }

  trackSoldQuantity(event: Trade, quantity: number): void {
    this.associateSoldTrade(event);

    const tracker = this.tradeSoldTracker.get(event.id);
    if (tracker === undefined) {
      throw new Error(`There is no tracker for the event (${event.id})`);
    }

    if (tracker.quantity + quantity > Math.abs(event.quantity)) {
      throw new Error(
        `Adding tracking quantity (${quantity}) to event (${event.id}) exceeds total quantity (${tracker.quantity + quantity} / ${Math.abs(event.quantity)}) of the event`,
      );
    }

    tracker.quantity += quantity;
  }

  getAcquiredTradeTracker(event: Trade): TradeAssociation {
    const tracker = this.tradeAcquiredTracker.get(event.id);
    if (tracker === undefined) {
      throw new Error(`Tracker for event (${event.id}) does not exist`);
    }
    return tracker;
  }

  getSoldTradeTracker(event: Trade): TradeAssociation {
    const tracker = this.tradeSoldTracker.get(event.id);
    if (tracker === undefined) {
      throw new Error(`Tracker for event (${event.id}) does not exist`);
    }
    return tracker;
  }
}
